cranelift/codegen/src/isa/x64/abi.rs

```diff
@@ -1,73 +1,602 @@
+            types::B16 | types::I16 => (true, 2),
+            types::B32 | types::I32 => (true, 4),
+            types::B64 | types::I64 | types::R64 => (true, 8),
+            types::F32 => (false, 4),
+            types::F64 => (false, 8),
+            _ => unimplemented!("store_stack({})", ty),
+        };
+
+        let mem = SyntheticAmode::from(mem);
+
+        if is_int {
+            Inst::mov_r_m(size, from_reg, mem, /* infallible store */ None)
+        } else {
+            let sse_op = match size {
+                4 => SseOpcode::Movss,
+                8 => SseOpcode::Movsd,
+                _ => unreachable!(),
+            };
+            Inst::xmm_mov_r_m(sse_op, from_reg, mem, /* infallible store */ None)
+        }
+    }
+
+    fn gen_move(to_reg: Writable<Reg>, from_reg: Reg, ty: Type) -> Self::I {
+        Inst::gen_move(to_reg, from_reg, ty)
+    }
+
+    /// Generate an integer-extend operation.
+    fn gen_extend(
+        to_reg: Writable<Reg>,
+        from_reg: Reg,
+        is_signed: bool,
+        from_bits: u8,
+        to_bits: u8,
+    ) -> Self::I {
+        let ext_mode = match from_bits {
+            1 | 8 => ExtMode::BQ,
+            16 => ExtMode::WQ,
+            32 => ExtMode::LQ,
+            _ => panic!("Bad extension: {} bits to {} bits", from_bits, to_bits),
+        };
+        if is_signed {
+            Inst::movsx_rm_r(ext_mode, RegMem::reg(from_reg), to_reg, None)
+        } else {
+            Inst::movzx_rm_r(ext_mode, RegMem::reg(from_reg), to_reg, None)
+        }
+    }
+
+    fn gen_ret() -> Self::I {
+        Inst::Ret
+    }
+
+    fn gen_epilogue_placeholder() -> Self::I {
+        Inst::EpiloguePlaceholder
+    }
+
+    fn gen_add_imm(into_reg: Writable<Reg>, from_reg: Reg, imm: u32) -> SmallVec<[Self::I; 4]> {
+        let mut ret = SmallVec::new();
+        if from_reg != into_reg.to_reg() {
+            ret.push(Inst::gen_move(into_reg, from_reg, I64));
+        }
+        ret.push(Inst::alu_rmi_r(
+            true,
+            AluRmiROpcode::Add,
+            RegMemImm::imm(imm),
+            into_reg,
+        ));
+        ret
+    }
+
+    fn gen_stack_lower_bound_trap(limit_reg: Reg) -> SmallVec<[Self::I; 2]> {
+        smallvec![
+            Inst::cmp_rmi_r(/* bytes = */ 8, RegMemImm::reg(regs::rsp()), limit_reg),
+            Inst::TrapIf {
+                // NBE == "> unsigned"; args above are reversed; this tests limit_reg > rsp.
+                cc: CC::NBE,
+                srcloc: SourceLoc::default(),
+                trap_code: TrapCode::StackOverflow,
+            },
+        ]
+    }
+
+    fn gen_get_stack_addr(mem: StackAMode, into_reg: Writable<Reg>, _ty: Type) -> Self::I {
+        let mem: SyntheticAmode = mem.into();
+        Inst::lea(mem, into_reg)
+    }
+
+    fn get_stacklimit_reg() -> Reg {
+        // As per comment on trait definition, we must return a caller-save
+        // register here.
+        regs::r10()
```

Comment:
Can we add an assertion that this is caller-save? (This depends on the calling convention, not sure if it's available from here.)

Reply:
Done; just asserting that it's a caller-save in systemv and baldrdash.